src/grt/src/GlobalRouter.cpp

```diff
@@ -2924,13 +2925,29 @@ int GlobalRouter::findObstructions(odb::Rect& die_area)
   return obstructions_cnt;
 }
 
+void GlobalRouter::extendObstructions(
+    std::unordered_map<int, odb::Rect>& macro_obs_per_layer,
+    int bottom_layer,
+    int top_layer)
+{
+  for (int l = top_layer; l > bottom_layer; l--) {
+    odb::Rect& layer_obs = macro_obs_per_layer[l - 1];
+    odb::Rect& upper_layer_obs = macro_obs_per_layer[l];
+    layer_obs.set_xlo(std::min(layer_obs.xMin(), upper_layer_obs.xMin()));
+    layer_obs.set_ylo(std::min(layer_obs.yMin(), upper_layer_obs.yMin()));
+    layer_obs.set_xhi(std::max(layer_obs.xMax(), upper_layer_obs.xMax()));
+    layer_obs.set_yhi(std::max(layer_obs.yMax(), upper_layer_obs.yMax()));
```

Comment:
This is rather conservative.  It ignores two cases

1) If you have multiple available layers you don't need to block them.  Imagine an M6 shelf would block M2-M5 in this code but that is unnecessary as there is plenty of available routing freedom.  I was thinking more of the case of a single layer that was blocked above and below (either by a blockage or because the layer is outside the routing layer range).

2) You only need to block signals running parallel to the edge (vertical layers along e&w; horizontal layers along N&S).  Otherwise you might be blocking access to pins.

Reply:
In the case where a layer is blocked above and below by other obstructions, how this layer obstruction should be extended?

Some scenarios:
- M3 (vertical) is aligned with M4, but M2 is slightly larger than M3. Should the M3 obstruction be extended to have the same width as the M2 obstruction?
- 
- M4 (horizontal) is aligned with M3, but M5 is slightly larger than M4. Should the M4 obstruction be extended to have the same height as the M5 obstruction?
- 
- M6 is slightly smaller than M5, but there is no available layer above it. Should M6 be extended to have the same width (or height) as M5?

Reply:
> * M3 (vertical) is aligned with M4, but M2 is slightly larger than M3. Should the M3 obstruction be extended to have the same width as the M2 obstruction?

no as above the m2 you have both m3 & m4 available
> * M4 (horizontal) is aligned with M3, but M5 is slightly larger than M4. Should the M4 obstruction be extended to have the same height as the M5 obstruction?

no as below m5 you have both m3 & m4 available
> * M6 is slightly smaller than M5, but there is no available layer above it. Should M6 be extended to have the same width (or height) as M5?

yes as m6 is block on both m5 and the above layer